### ISSUE TITLE
VRTK Core Interaction Refactor

### ISSUE THREAD

**Issue by dustinlacewell** [thestonefox/VRTK#875]
> This proposal has been co-authored by: @dustinlacewell and @bddckr 

# tl;dr
-  Move event propagation from `ControllerEvents.Update` to `ControllerSDK.Update`
-  Replace `ControllerEvents` and `ControllerActions` with `ControllerBridge`
-  Deprecate `ControllerEvents`, `ControllerActions` and the 'script alias' fields on the `SDKManager`
-  Introduce `SDKManager` fields to represent the "current" bridges for left/right/etc controller devices
-  Propagate events from `ControllerSDK.Update` to the currently assigned bridges via the new fields
-  Perform tracking upon `ControllerBridge.trackedTransform` in `SDKManager.Update` versus re-parenting on startup.

# Summary
The idea is to do event generation, once, in `ControllerSDK` and propagate them to the relevant `ControllerBridge`s currently assigned to the `SDKManager`. This is both a bit more efficient, but also allows for an elegant mechanism of swapping around whole controller interaction mechanisms.

> Standard interaction scripts, such as `InteractTouch`, `InteractGrab`, `InteractUse` and `InteractableObject` are not the topic. The existing model of interaction they implement should continue to work.

# How VRTK controller events work

In VRTK, `ControllerEvent` components poll the `SDK_Bridge` for controller state and emit events for the benefit of subscribers, usually interaction components. Interaction components may also utilize `ControllerAction` components as an API for initiating haptics and manipulating the controller mesh.

# How VRTK interactions work

> For simplicity's sake, despite there usually being two tracked controllers they'll be talked about as one.

As described, `ControllerEvent` components emit events to interaction scripts so they can implement their behaviours. This `ControllerEvent` instance is placed on a `GameObject` along with any of those interaction scripts.

Many interaction scripts use the 'script alias' field on the `SDKManager` to get a reference to this GO where all the scripts for a controller live. This alias is set when play begins. Changing it during run-time will change which `ControllerEvents` instance code that uses these fields will find. However, VRTK does not currently use these fields for anything internally, except for transform tracking which is described next.

# How VRTK tracking works

When Play-mode begins, VRTK utilizes the `SDKManager` to identify the current `GameObject` assigned as the 'script alias'. It then relocates this `GameObject` to be parented to the SDK tracked controller `GameObject`. As the SDK tracked controller GO is transformed, so is the now relocated "script alias" GO.

# Identified problems with VRTK

- Changing the script alias does not update tracking
- Changing the script alias does not affect event propagation
- Any number of `ControllerEvents` may poll and generate events
- Using the `ControllerEvents` parent GO for tracking is limiting
- Privacy of fields connecting interaction scripts to `ControllerEvents` is limiting

Each user must handle the above complexity by either:
- twiddling settings on single instances of interaction scripts to model changes in user interactivity
- emulating the swapability provided by this proposal

# Proposed Changes

*Minor Changes*:
  - Make `public` the fields of interaction scripts that depend core interaction scripts

*Major Changes*:

  - Merge `ControllerEvents` and `ControllerActions` into `ControllerBridge`
  - Deprecate all related 'script alias' fields and helper methods from `SDKManager`
  - Create `SDKManager` fields to refer to the current bridge for each controller device
  - Track `ControllerBridge.trackedTransform` instead of its parent GO.
  - Move event generation from `ControllerEvents` to `ControllerSDK.Update`
  - Only emit events on currently assigned bridges on the `SDKManager`

*Related Changes*:
  - Rethink event virtualization

## Public Dependency Fields

Making `public`  the fields which link core interaction scripts to the event and action components they no longer have the artificial limitation of being required to exist on the same object. This allows for some flexibility by linking up these dependencies in the inspector or programatically without inheritance. 

Currently these private fields are automatically assigned on script startup. This feature can be reserved in the case that the now public fields are left `null`.

## ControllerBridge

The merging of the existing `ControllerEvents` and `ControllerActions`, it is the single interface for "user components" to interact with individual SDK-managed devices. 

### Event Emission

The `ControllerBridge` works much like before. It has events for "user components" interested in device events. Interaction scripts will still register handlers in the way they did before. However, given that their dependency fields are no longer public, these scripts will not be required to exist on the same GO as the bridge.

### Actions

Just as before, haptics and highlighting of the device associated with the bridge will be possible. Simply put, the API for `ControllerActions` has been absorbed by `ControllerBridge`.

### Tracked Transform

A new field, `ControllerBridge.trackedTransform` will identify the the transform that should be translated to match the tracking of the associated SDK device. This frees the `ControllerBridge` from having to specifically be located upon the tracked `GameObject`.

## SDKManager Fields

At least two new fields are being introduced to the `SDKManager`:

- `ControllerBridge SDKManager.leftBridge`;
- `ControllerBridge SDKManager.rightBridge`;

These fields replace the "script alias" fields. By tracking `ControllerBridge` components directly, the dependency on their parent `GameObject`s is further reduced, such in the case of tracking. 

## Transform Tracking

*Old way*:  Reparent the GO assigned as the "script alias" on startup to the SDK-tracked GO.
*New way*: `ControllerSDK` uses the `SDKManager`'s bridge fields to get at `ControllerBridge.trackedTransform`

## Event Propagation

Event propagation will be changed from polling in the old `ControllerEvents` component against the SDK, to the `ControllerSDK.Update` method. Events will be sent directly the the relevant currently assigned `ControllerBridge` on the `SDKManager`.

This creates one of the largest benefits of the bridge architecture. By simply reassigning the bridge fields on the `SDKManager` the flow of events can be altered. Every `ControllerBridge` acts as the root event source for a tree of interdependent interaction scripts. This indirection allows for dynamically controlling everything about a players interactability with the game. All without having any arbitrary or artificial constraints regarding the structure of GameObjects, the proximity of interaction components, or any of the previously mentioned concerns.

**Comment by thestonefox** [comment]
This needs a tldr; and probably a nice diagram :)

**Comment by dustinlacewell** [comment]
@thestonefox @bddckr The issue has been completed and its wordiness reduced. 

@thestonefox I think now you should be able to get where we're going without a diagram. (Which I attempted and then lost patience for 😄 )

**Comment by thestonefox** [comment]
I'm going to start putting some blurb down to describe (my thoughts) of what the current system is intending to solve and what actually is the theory behind the problem that needs solving. It's not a reference to any previous proposals just a brain dump from myself to try and help the discussion (it may be already very aligned with other thoughts anyway).

Also, for reference, this issue may be relevant https://github.com/thestonefox/VRTK/issues/842

Ok, so brain dump time!!

The scripts that care about a controller are:

 * Controller Events
 * Controller Actions
 * Interact Touch
 * Interact Grab
 * Interact Use
 * Object Auto Grab
 * Pointers (simple/bezier)
 * UI Pointer

# Controller Events

tldr; I think Controller Events should be deprecated and replaced with something the user doesn't even have to add to a GameObject, it's just a known thing.

### Concept

The concept of controller events is to poll the button presses on the VR SDK and when a press is valid, it then emits an event with that button action.

The controller events also handle the concept of a global button alias (such as pointer activate, grab, use, etc)


### Thoughts

My thoughts are firstly, do we even need to poll these SDK controller states to raise events? Do the SDKs themselves not emit an event we can listen into and then have the SDK Bridge for the controller just abstract that event into being a standard VRTK event?

All that really is needed is a common event emitted when a specific button is pressed on a controller and then anything interested in button presses subscribes to that event.

The only problem with doing it on the SDK Bridge for Controller is you need to know "which" controller you're interested in subscribing to (perhaps based on index if we can agree all controllers including pucks need a unique identifier).

I think having the aliases in the `Controller Events` as a global thing is wrong and stupid. Controller Events shouldn't need to know about these aliases, it should (or whatever replaces it) should just know about the controller button being actioned and throw the relevant event. Then the sub scripts should just store the button that activates them (e.g. Pointers just have an activation button like touchpad control does).

If whatever raises these events sends out an event for a button press and the controller index, then we need some way to tell other scripts you're only interested in registering on this controller index for your event (and potentially switching the listener if you swap controllers).


# Controller Actions

tldr; Controller Actions should be split up into more specific scripts so it's not just a generic dumping ground, the smaller, concise scripts have to have some knowledge about the controller model sub game objects.

### Concept

These deal with doing something on the controller such as vibrating it, highlighting it or dealing with the elements that make up the controller (e.g. changing their opacity).

### Thoughts

I think this is probably a candidate again for something that can be split up into smaller scripts that don't need to do so much.

Vibrations can be abstracted into the same level as the new Controller Events stuff, you just need to know an index to call the SDK vibration stuff.

The highlighting and opacity stuff needs some way of knowing the paths to the model elements and give you a way of overriding them like it currently does but that can probably be handled in a separate script. One that deals with the controller element appearance. At the moment it's too tied into how the vive works and isn't abstracted enough.


# Interact Touch

tldr; Interact touch is too heavily coupled at the moment to a controller and to interact grab/use. All it needs to be is a script that sits on the actual controller game obejct listening for the collision enter/exits and emit an event.

### Concept

The interact touch creates a collider that has to track absolutely where the current controller game object is, this is why it becomes a child of the actual game object. Just tracking the controller game object's transform (position/rotation/scale) doesn't give a high enough level of fidelity. 

### Thoughts

So Interact Touch is very coupled at the moment to Interact Grab and Interact Use, which is daft. Really what should happen is Interact Touch is added to the controller game object. All it does is adds the trigger collider and rigidbody to track the controller and emits an event on touch and on stop touching. You're also able to override the given game object that contains the custom collider.

Interact Touch needs linking to a controller in some way, at the moment it's done by the fact that you have to put Interact Touch on the same component as Controller Events, but then maybe if in the SDK Manager (or somewhere) you just specify an array of controllers, then interact touch just has a parameter where you select which controller you're interested in tracking for touches.


# Interact Grab

tldr; Interact grab should be concerned with it's own button alias for grabbing (not in controller events). It shouldn't be coupled to any given script and should just know which controller it should listen to events for (e.g. controller touched, controller button pressed for grab alias).

### Concept

Interact grab deals with attaching a valid touched object to a given point on the "grabbing" controller.

### Thoughts

Currently, the Interact Grab script gets a direct reference to the Interact Touch script that has to be sitting on the same Game Object. This is unnecessary as all that needs to happen is the interact grab script just needs to subscribe to the `touch` / `untouch` event of a specific controller (based on index as mentioned above).

Once the event comes through that an object is being touched, then Interact Grab can store a reference to it waiting for the grab script to be pressed. Any info required about the controller, such as the attach point can be looked up via whatever controller description mechanism there is (e.g. calling the controller SDK bridge asking for the attach point if one isn't provided).

Also, grab uses a "Grab Alias" which is currently on the controller events. But that alias could simply live on the interact grab script. Then switching the grab alias could be a case of just amending the parameters on this grab script. It just listens for the selected alias event.

The other thing grab does, is turns the rigidbody on the controller to non-kinematic. But this could (probably should) be handled by the interact touch (which is responsible for the controller collider/rigidbody) can subscribe to an event that toggles the rigidbody and the interact grab (plus anything else) just emits that event.

Interactable Objects also deal with grabbing where it can deny a certain controller from being able to grab the object, but then on the grab attempt it can simply check that the grabbed object is not restricted from being grabbed by the controller that interact grab is subscribed to.


# Interact Use

tldr; Pretty much the same as interact grab

### Concept

Interact Use is similar to interact grab, it listens for an alias and then calls the using action on the object being touched.

### Thoughts

As the same as interact grab it's way too coupled at the moment. It just needs to have the same concept as grab and listen for a touch event and then register the touched object to deal with using if the use button is pressed. The use button should be localised to the use script and not globally held elsewhere.


# Object Auto Grab

tldr; Way too coupled at the moment, just needs references.

### Concept

To automatically snap an object to the grabbed position on the controller the script is sitting on.

### Thoughts

This is far too coupled to needing to be on the controller script. Again it should be able to go anywhere and just have a reference to the controller to snap to and the object to snap. It should be very simple and decoupled.


# Pointers (Simple/Bezier)

tldr; Pointers are mostly decoupled at the moment, but still have some legacy dependencies like needing a global button alias when they can be responsible for their own. They also need to know about the controller if they want to interact with objects to add a collider.

### Concept

They emit a beam of a certain shape and reference a point in world space that can be used for something (e.g. teleport to that location).

### Thoughts

Pointers are already largely de-coupled from the controller as you give it a controller game object for it to listen to the button presses on. However, if you dont give it a game object for a controller then it looks at the scripts on the current game object.

If a pointer just works in the same way the other scripts above work, where they reference a controller to listen to the button events on and the pointer is also in control of setting it's own pointer activate and pointer set button alias, then this should be relatively decoupled.

Pointers can already exist on game objects that aren't the controller, they just need to know which controller is turning them on or off.

Pointers do have the ability to extend a controller's collider set though to allow for interacting with objects via a pointer. Because of this they do still need to know which controller they are a part of so they can create the relevant child game objects of that controller for touch tracking.


# UI Pointer

This is pretty much the same as the pointer. Anything that works on the pointer should be fine here too. It's just a button alias (that the UI Pointer should own) and a ray cast from a given origin (e.g. tell it the controller to raycast from).



---------------------------

That's all I have so far, I'll probably look to add more to this when I think of more. I may also be wrong, so any questions or discussion points then please ask!

**Comment by bddckr** [comment]
@thestonefox I agree about the whole "button alias settings should be on the specific objects that use them, not somewhere else". I think the idea of "virtual events" should be able to solve that whole problem:
- SDKs emit their own SDK-specific events. If the vendor SDKs don't have events we just poll in the ControllerSDK implementation and emit them there.
- Users create a "virtual event" and set which SDK-specific event they want to listen to to emit this "virtual event".
- For events like axis changes it should be possible to hook some kind of "event validator" between the SDK-specific event and the "virtual event".
Example: "Virtual event" called "Fire gun". Uses trigger axis event for Vive on SteamVR and same on Rift with Oculus' SDK. There is an event validator that checks the axis for something like "is >= 0.9" or something.

I really think this and #842 should move the button mapping stuff into the SDKs. It's already a pain having the stupid "Button One/Two" in VRTK (they aren't descriptive at all!) and we're getting support for new SDKs added currently.

**Comment by thestonefox** [comment]
@bddckr When you say "User creates a virtual event" what do you mean by this?

A new script called "virtual event" that you hook up to a controller index then it throws an event when the given button is pressed on the given controller?

**Comment by bddckr** [comment]
@thestonefox Virtual Event is just a C# class. The user creates them via a list on ControllerEvents or somewhere else. That list is then used to let the SDKs let the Virtual Events know about the emitted SDK Event. Or the Virtual Events subscribe directly to the specified SDK Events (with possible Event Validator in between).

**Comment by thestonefox** [comment]
Merging with https://github.com/thestonefox/VRTK/issues/1523
